src/tests.rs

```diff
@@ -94,3 +96,181 @@ async fn test_disputes_offchain_disabling() -> Result<(), Error> {
+    let version_before = get_runtime_version(&client).await?;
+
+    // runtime upgrade
+    // Put a runtime from the disabling branch compiled with 'fast runtime'
+    // as artifacts/westend_runtime.compact.compressed.wasm
+    // compile command: `cargo build --release --features=fast-runtime -p westend-runtime`
+    let code = std::fs::read(
+        Path::new(env!("CARGO_MANIFEST_DIR"))
+            .join("artifacts/westend_runtime.compact.compressed.wasm"),
+    )?;
+
+    println!("Starting runtime upgrade");
+    perform_runtime_upgrade(&client, code).await?;
+    println!("🚀🚀🚀 runtime upgrade complete");
+
+    assert_blocks_are_being_finalized(&client).await?;
+
+    // get runtime version after the upgrade - is it newer?
+    let version_after = get_runtime_version(&client).await?;
+    assert!(version_after.spec_version > version_before.spec_version);
+
+    // still no disputes
+    wait_for_metric(honest, DISPUTES_CONCLUDED_VALID, 0).await?;
+
+    Ok(())
+}
+
+// All prereqs for `test_runtime_upgrade` are valid here too.
+// The test performs runtime upgrade with old client nodes. They should work fine with the new runtime.
+#[tokio::test]
+async fn test_runtime_upgrade_with_old_client() -> Result<(), Error> {
+    tracing_subscriber::fmt::init();
+
+    let network = spawn_honest_network().await?;
+
+    println!("🚀🚀🚀 network deployed");
+
+    let honest = network.get_node("honest-0")?;
+    let role = honest.reports("node_roles").await?;
+    assert_eq!(role as u64, 4);
+
+    // No disputes in honest network
+    wait_for_metric(honest, DISPUTES_CONCLUDED_VALID, 0).await?;
+
+    let client = get_client(&network, "honest-0").await?;
+
+    // Netowrk is still healthy
+    assert_blocks_are_being_finalized(&client).await?;
+
+    // get runtime version before the upgrade
+    let version_before = get_runtime_version(&client).await?;
+
+    // runtime upgrade
+    let code = std::fs::read(
+        Path::new(env!("CARGO_MANIFEST_DIR"))
+            .join("artifacts/westend_runtime.compact.compressed.wasm"),
+    )?;
+
+    println!("Starting runtime upgrade");
+    perform_runtime_upgrade(&client, code).await?;
+    println!("🚀🚀🚀 runtime upgrade complete");
+
+    assert_blocks_are_being_finalized(&client).await?;
+
+    // get runtime version after the upgrade - is it newer?
+    let version_after = get_runtime_version(&client).await?;
+    assert!(version_after.spec_version > version_before.spec_version);
+
+    // still no disputes
+    wait_for_metric(honest, DISPUTES_CONCLUDED_VALID, 0).await?;
+
+    Ok(())
+}
+
+async fn assert_blocks_are_being_finalized(
+    client: &OnlineClient<PolkadotConfig>,
+) -> Result<(), Error> {
+    let mut finalized_blocks = client.blocks().subscribe_finalized().await?;
+    let first_measurement = finalized_blocks
+        .next()
+        .await
+        .ok_or(Error::from("Can't get finalized block from stream"))??
+        .number();
+    sleep(Duration::from_secs(30)).await;
```

Comment:
not a fan of sleeps for asserting progress, this is the reason why tests are so flaky in CI
i'd rather have one timeout on the whole test, cf https://github.com/paritytech/zombienet-sdk/pull/145#issuecomment-1868545533

also, this sleep is not acting as a timeout since we `.await` later anyway
```suggestion
```

Reply:
I agree. The sleep is pointless here. Initially I wanted to verify the network is producing blocks (30 sec, ~5 blocks), then decided it's an overkill, simplified the assert and forgot to remove the sleep :dizzy_face: